src/charm.py

```diff
@@ -105,6 +110,11 @@ def _on_leader_elected(self, event) -> None:
 
         self._update_hosts(event)
 
+        if "db_initialised" not in self.app_data:
+            return
+        time.sleep(MEDIAN_REELECTION_TIME * 2)
+        self.update_app_relation_data()
```

Comment:
I don't know what I was thinking when I suggested this, `event.defer()` would be much better here
```suggestion
try:
       self.update_app_relation_data(check_mongo_users=True)
except OperationFailure as e:
       # when a unit departs the replica set it is possible that unit was a leader, thus resulting in mongod electing a new leader. If mongod hasn't elected a 
       # new leader yet an Authentication error will occur and we should retry updating the application relation data.
       if e.code == 18:
             logger.debug("Waiting for replicaset to elect a new primary before updating relation data")
             event.defer()
       else:
              raise
```